### PR TITLE
Add nondeterministic alert for `index_put_` when `accumulate=False`

### LINKED ISSUE #55516: Add nondeterministic alert to `index_put_`
`index_put_` is nondeterministic when given duplicate indices when `accumulate=False`

Related: #50750

cc @mruberry @kurtamohler

### PR DESCRIPTION
**Pull request by kurtamohler**
Fixes #55516

### CODE REVIEW

**Review by ngimel — APPROVED** [review]
Thank you for the fix

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp [hunk 1]
```diff
@@ -434,6 +434,11 @@ Tensor take(const Tensor& self, const Tensor& index) {
 }
 
 Tensor & index_put_(Tensor & self, const torch::List<c10::optional<Tensor>>& indices, const Tensor & value, const bool accumulate) {
+  if (!accumulate) {
```
Comment (Lezcano) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:437]: Can you check whether it's deterministic on GPU? In `put_` I was using `atomicGPUAdd`, which is apparently non-deterministic, so I had to write:
https://github.com/pytorch/pytorch/blob/7d49c82cc545b6839387e7cc53104af0cf4b9594/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L321-L324

Comment (kurtamohler) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:437]: Good idea! I looked through the code for `index_put_` CUDA with `accumulate=True` (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Indexing.cu#L187), and I don't see anything that is obviously nondeterministic

Comment (ngimel) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:437]: index_put_ with accum=True is deterministic on cuda

Comment (kurtamohler) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:437]: Great, thanks for the confirmation @ngimel !

**Pull request closed without merging**